Intersections_3/include/CGAL/Intersections_3/internal/tetrahedron_lines_intersections_3.h

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2019 GeometryFactory(France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org); you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public License as
+// published by the Free Software Foundation; either version 3 of the License,
+// or (at your option) any later version.
+//
+// Licensees holding a valid commercial license may use this file in
+// accordance with the commercial license agreement provided with the software.
+//
+// This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+// WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: LGPL-3.0+
+//
+//
+// Author(s)     : Maxime Gimeno
+//
+
+#ifndef TETRAHEDRON_LINES_INTERSECTIONS_3_H
+#define TETRAHEDRON_LINES_INTERSECTIONS_3_H
+
+#include <vector>
+#include <CGAL/kernel_basic.h>
+#include <CGAL/intersections.h>
+
+namespace CGAL {
+
+namespace Intersections {
+
+namespace internal {
+template<class K, class O, class T>
+struct Tetrahedron_lines_intersection_3_base
+{
+  typedef typename Intersection_traits<K,
+      CGAL::Tetrahedron_3<K>,
+      O>::result_type Result_type;
+
+  typedef typename Intersection_traits<K,
+      CGAL::Triangle_3<K>,
+      O>::result_type Inter_type;
+
+  Tetrahedron_lines_intersection_3_base(const typename K::Tetrahedron_3& tet,
+  const O& o):tet(tet), o(o)
+  {}
+  const typename K::Tetrahedron_3& tet;
+  const O& o;
+  std::vector<typename K::Point_3> res_points;
+  Result_type output;
+  bool all_inside_test()
+  {
+    return false;
+  }
+
+  bool are_extremities_inside_test()
+  {
+    return false;
+  }
+
+  void do_procede()
+  {
+    if(static_cast<T*>(this)->all_inside_test())
+      return;
+
+    int res_id = -1;
+
+    Inter_type tr_seg[4];
+    for(std::size_t i = 0; i < 4; ++i)
+    {
+     const typename K::Triangle_3 triangle(tet.vertex((i+1)%4),
+                             tet.vertex((i+2)%4),
+                             tet.vertex((i+3)%4));
+      tr_seg[i] = typename K::Intersect_3()(o, triangle);
```

Comment:
Same problem as at other places, you should use calls to do_intersect to deduce your result before calling actual constructions.